### PR TITLE
Port to big-endian platforms (s390x but others can be trivially added)

### LINKED ISSUE #1063: 247 test failures on big-endian arches (ppc64 and s390x)
Running the 1.0.10 testsuite (test/run-tests.py) on big-endian arches (POWER8/ppc64 and s390x) yields the following failed tests:
```

- test/binary/bad-call-indirect-reserved.txt
- test/binary/bad-data-count-mismatch.txt
- test/binary/bad-data-count-order-after-code.txt
- test/binary/bad-data-count-order-before-elem.txt
- test/binary/bad-data-size.txt
- test/binary/bad-duplicate-section-around-custom.txt
- test/binary/bad-duplicate-section.txt
- test/binary/bad-duplicate-subsection.txt
- test/binary/bad-event-after-export.txt
- test/binary/bad-event-before-global.txt
- test/binary/bad-export-func.txt
- test/binary/bad-extra-end.txt
- test/binary/bad-function-body-count.txt
- test/binary/bad-function-body-size.txt
- test/binary/bad-function-count-missing-code-section.txt
- test/binary/bad-function-local-count.txt
- test/binary/bad-function-local-type.txt
- test/binary/bad-function-names-too-many.txt
- test/binary/bad-function-param-type.txt
- test/binary/bad-function-result-type.txt
- test/binary/bad-function-sig.txt
- test/binary/bad-function-too-many-results.txt
- test/binary/bad-import-sig.txt
- test/binary/bad-linking-metadata.txt
- test/binary/bad-logging-basic.txt
- test/binary/bad-memory-grow-reserved.txt
- test/binary/bad-memory-init-max-size.txt
- test/binary/bad-memory-init-size.txt
- test/binary/bad-memory-max-size.txt
- test/binary/bad-memory-size-reserved.txt
- test/binary/bad-name-section-invalid-index.txt
- test/binary/bad-name-section-location.txt
- test/binary/bad-names-duplicate-locals.txt
- test/binary/bad-names-duplicates.txt
- test/binary/bad-names-function-locals-out-of-order.txt
- test/binary/bad-names-out-of-order.txt
- test/binary/bad-op-after-end.txt
- test/binary/bad-opcode-prefix.txt
- test/binary/bad-opcode.txt
- test/binary/bad-returncallindirect-reserved.txt
- test/binary/bad-section-ends-early.txt
- test/binary/bad-section-size-zero.txt
- test/binary/bad-segment-no-memory.txt
- test/binary/bad-simd-type.txt
- test/binary/bad-start-func.txt
- test/binary/bad-subsection-out-of-order.txt
- test/binary/bad-subsection-size.txt
- test/binary/bad-subsection-unfinished.txt
- test/binary/bad-too-many-locals.txt
- test/binary/bad-type-form.txt
- test/binary/bad-typecheck-fail.txt
- test/binary/bad-typecheck-missing-drop.txt
- test/binary/bad-version.txt
- test/binary/basic.txt
- test/binary/duplicate-func-names.txt
- test/binary/duplicate-local-names.txt
- test/binary/dylink-section.txt
- test/binary/ignore-custom-section-error-objdump.txt
- test/binary/ignore-custom-section-error-wasm2wat.txt
- test/binary/invalid-name.txt
- test/binary/linking-section.txt
- test/binary/missing-code-section-empty-function-section.txt
- test/binary/missing-function-section-empty-code-section.txt
- test/binary/names.txt
- test/binary/no-global-names.txt
- test/binary/no-names.txt
- test/binary/relocs.txt
- test/binary/user-section.txt
- test/dump/bad-version-logging.txt
- test/dump/bad-version.txt
- test/dump/basic.txt
- test/dump/binary.txt
- test/dump/block-257-exprs-br.txt
- test/dump/block-257-exprs.txt
- test/dump/block-multi.txt
- test/dump/block.txt
- test/dump/br-block-named.txt
- test/dump/br-block.txt
- test/dump/br-loop-inner-expr.txt
- test/dump/br-loop-inner.txt
- test/dump/br-loop.txt
- test/dump/br_on_exn.txt
- test/dump/brif-loop.txt
- test/dump/brif.txt
- test/dump/brtable-empty.txt
- test/dump/brtable.txt
- test/dump/call.txt
- test/dump/callimport.txt
- test/dump/callindirect.txt
- test/dump/cast.txt
- test/dump/compare.txt
- test/dump/const.txt
- test/dump/convert-sat.txt
- test/dump/convert.txt
- test/dump/current-memory.txt
- test/dump/debug-import-names.txt
- test/dump/debug-names.txt
- test/dump/dedupe-sig.txt
- test/dump/drop.txt
- test/dump/event.txt
- test/dump/export-multi.txt
- test/dump/expr-br.txt
- test/dump/expr-brif.txt
- test/dump/func-exported.txt
- test/dump/func-multi.txt
- test/dump/func-named.txt
- test/dump/func-result-multi.txt
- test/dump/getglobal.txt
- test/dump/getlocal-param.txt
- test/dump/getlocal.txt
- test/dump/global.txt
- test/dump/grow-memory.txt
- test/dump/hexfloat_f32.txt
- test/dump/hexfloat_f64.txt
- test/dump/if-multi.txt
- test/dump/if-then-else-list.txt
- test/dump/if-then-list.txt
- test/dump/if.txt
- test/dump/import.txt
- test/dump/invalid-data-segment-no-memory.txt
- test/dump/invalid-data-segment-offset.txt
- test/dump/invalid-elem-segment-no-table.txt
- test/dump/invalid-elem-segment-offset.txt
- test/dump/load-aligned.txt
- test/dump/load.txt
- test/dump/locals.txt
- test/dump/loop-257-exprs-br.txt
- test/dump/loop-257-exprs.txt
- test/dump/loop-multi.txt
- test/dump/loop.txt
- test/dump/memory-1-byte.txt
- test/dump/memory-data-size.txt
- test/dump/memory-hex.txt
- test/dump/memory.txt
- test/dump/module-name.txt
- test/dump/mutable-global.txt
- test/dump/no-canonicalize.txt
- test/dump/nocheck.txt
- test/dump/nop.txt
- test/dump/param-multi.txt
- test/dump/result.txt
- test/dump/rethrow.txt
- test/dump/return.txt
- test/dump/select.txt
- test/dump/setglobal.txt
- test/dump/setlocal-param.txt
- test/dump/setlocal.txt
- test/dump/signatures.txt
- test/dump/simd-basic.txt
- test/dump/simd-binary.txt
- test/dump/simd-bitselect.txt
- test/dump/simd-compare.txt
- test/dump/simd-lane.txt
- test/dump/simd-load-store.txt
- test/dump/simd-shift.txt
- test/dump/simd-splat.txt
- test/dump/simd-unary.txt
- test/dump/start.txt
- test/dump/store-aligned.txt
- test/dump/store.txt
- test/dump/string-escape.txt
- test/dump/string-hex.txt
- test/dump/table.txt
- test/dump/tee_local.txt
- test/dump/throw.txt
- test/dump/try-multi.txt
- test/dump/try.txt
- test/dump/unary-extend.txt
- test/dump/unary.txt
- test/dump/unreachable.txt
- test/gen-spec-js/action.txt
- test/gen-spec-js/assert_exhaustion.txt
- test/gen-spec-js/assert_return.txt
- test/gen-spec-js/assert_return_nan.txt
- test/gen-spec-js/assert_trap.txt
- test/gen-spec-js/assert_uninstantiable.txt
- test/gen-spec-js/assert_unlinkable.txt
- test/gen-spec-js/basic.txt
- test/gen-spec-js/many-modules.txt
- test/gen-spec-js/register.txt
- test/interp/atomic-load.txt
- test/interp/atomic-rmw-add.txt
- test/interp/atomic-rmw-and.txt
- test/interp/atomic-rmw-cmpxchg.txt
- test/interp/atomic-rmw-or.txt
- test/interp/atomic-rmw-sub.txt
- test/interp/atomic-rmw-xchg.txt
- test/interp/atomic-rmw-xor.txt
- test/interp/atomic-store.txt
- test/interp/basic-logging.txt
- test/interp/load.txt
- test/interp/logging-all-opcodes.txt
- test/interp/simd-binary.txt
- test/interp/simd-compare.txt
- test/interp/simd-lane.txt
- test/interp/simd-load-store.txt
- test/interp/simd-shift.txt
- test/interp/simd-splat.txt
- test/interp/simd-unary.txt
- test/interp/store.txt
- test/opcodecnt/basic.txt
- test/parse/expr/bad-simd-shuffle-lane-index-overflow.txt
- test/parse/module/binary-module.txt
- test/regress/regress-20.txt
- test/regress/regress-24.txt
- test/regress/regress-7.txt
- test/regress/regress-8.txt
- test/spec/address.txt
- test/spec/binary.txt
- test/spec/custom.txt
- test/spec/endianness.txt
- test/spec/float_exprs.txt
- test/spec/float_literals.txt
- test/spec/float_memory.txt
- test/spec/globals.txt
- test/spec/imports.txt
- test/spec/left-to-right.txt
- test/spec/memory.txt
- test/spec/memory_redundancy.txt
- test/spec/memory_trap.txt
- test/spec/mutable-global/globals.txt
- test/spec/utf8-custom-section-id.txt
- test/spec/utf8-import-field.txt
- test/spec/utf8-import-module.txt
- test/strip/basic.txt
- test/strip/names.txt
- test/strip/no-custom-sections.txt
- test/two-commands.txt
- test/wasm2c/spec/address.txt
- test/wasm2c/spec/binary.txt
- test/wasm2c/spec/custom.txt
- test/wasm2c/spec/endianness.txt
- test/wasm2c/spec/float_exprs.txt
- test/wasm2c/spec/float_literals.txt
- test/wasm2c/spec/float_memory.txt
- test/wasm2c/spec/imports.txt
- test/wasm2c/spec/left-to-right.txt
- test/wasm2c/spec/memory.txt
- test/wasm2c/spec/memory_redundancy.txt
- test/wasm2c/spec/memory_trap.txt
- test/wasm2c/spec/token.txt
- test/wasm2c/spec/typecheck.txt
- test/wasm2c/spec/unreached-invalid.txt
- test/wasm2c/spec/utf8-custom-section-id.txt
- test/wasm2c/spec/utf8-import-field.txt
- test/wasm2c/spec/utf8-import-module.txt
- test/wasm2c/spec/utf8-invalid-encoding.txt
```

Is running on big-endian supported at all? Only 1027 tests passed and 1702 were skipped, so this is a significant portion.

### PR DESCRIPTION
**Pull request by SoniEx2**
closes #1063

~~61~~ ~~59~~ ~~40~~ ~~22~~ ~~13~~ ~~10~~ ~~7~~ 1 failing tests https://gist.github.com/SoniEx2/4379beb5909d0d39d52b48a53b7247b7

This ports wabt to big-endian platforms, and also ports the output of wasm2c to big-endian platforms (for cross-compilation needs). It's not specific to s390x other than the makefile changes.

Additionally the following unittests and c-api-tests fail, as they interact with wasm memory but haven't been adjusted to take memory ordering into account (and we think that should be a separate pull request and/or blocked on further specs like e.g. https://github.com/WebAssembly/wasm-c-api/issues/156 ):

```
unittests:
/home/linux1/wabt/src/test-interp.cc:545: Failure
Expected equality of these values:
  "Uryyb, JroNffrzoyl!"
  string_data
    Which is: "Hello, WebAssembly!"
[  FAILED  ] InterpTest.Rot13 (0 ms)


c-api-tests:
Running.. /home/linux1/wabt/out/wasm-c-api-memory
Failed with returncode=1
b'CAPI: [wasm_byte_vec_new_uninitialized] 92\nCAPI: [wasm_byte_vec_copy] 92\nCAPI: [wasm_byte_vec_new_uninitialized] 92\nCAPI: [wasm_byte_vec_delete] \nCAPI: [wasm_instance_new] 0x1e5beee0 0x1e5c0650\nCAPI: [wasm_extern_vec_new_uninitialized] 4\nCAPI: [wasm_instance_exports] 4\nCAPI: [wasm_module_delete] \nCAPI: [wasm_byte_vec_delete] \nCAPI: [wasm_memory_copy] \nCAPI: [wasm_memory_same] \nCAPI: [wasm_memory_delete] \nInitializing...\nLoading binary...\nCompiling module...\nInstantiating module...\nExtracting exports...\nChecking memory...\n> Error, expected success\n'
FAIL(1): /home/linux1/wabt/out/wasm-c-api-memory
```

Other than these, everything else passes.

### CODE REVIEW

**Comment by SoniEx2** [comment]
the gen-spec-js failures seem to be unrelated. not sure what to do about it.

File: src/interp/interp.cc [hunk 1]
```diff
@@ -531,10 +531,13 @@ Result Memory::Match(class Store& store,
...
   u64 new_pages;
   if (CanGrow<u64>(type_.limits, pages_, count, &new_pages)) {
+#if WABT_BIG_ENDIAN
+    auto old_size = data_.size();
+#endif
     pages_ = new_pages;
     data_.resize(new_pages * WABT_PAGE_SIZE);
 #if WABT_BIG_ENDIAN
-    /* FIXME move data around */
+    std::move_backward(data_.begin(), data_.begin() + old_size, data_.end());
```
Comment (SoniEx2) [inline src/interp/interp.cc:540]: this is likely wrong, but no tests are failing because of it. shouldn't this be filling the space with NULs or something? this might need a new test...

**Comment by SoniEx2** [comment]
we're gonna just ignore that one failing test and pretend it doesn't exist

File: src/common.h [hunk 2]
```diff
@@ -140,15 +140,23 @@ struct v128 {
     static_assert(sizeof(T) <= sizeof(v), "Invalid cast!");
     assert((lane + 1) * sizeof(T) <= sizeof(v));
     T result;
+#if WABT_BIG_ENDIAN
```
Comment (binji) [inline src/common.h:143]: Refactor this into shared code?

File: src/interp/interp.cc [hunk 3]
```diff
@@ -552,9 +563,21 @@ Result Memory::Init(u64 dst_offset,
                     u64 size) {
   if (IsValidAccess(dst_offset, 0, size) &&
       src.IsValidRange(src_offset, size)) {
+#if WABT_BIG_ENDIAN
+    std::copy(src.desc().data.begin() + src_offset,
+              src.desc().data.begin() + src_offset + size,
+              data_.end() - dst_offset - size);
+    auto base = data_.end() - dst_offset - size;
+    for (size_t i = 0; i < (size>>1); i++) {
```
Comment (binji) [inline src/interp/interp.cc:571]: Refactor into `SwapBytesInRange` (or better name?) function

**Review by binji — COMMENTED** [review]
Thanks for working on this! In general, looks pretty good. I'm mostly concerned about the code duplication, and that it is a bit hard to tell at first that big-endian stores all the data backward. It might be nice to refactor this into a set of common functions for operating on memory, and add some documentation making this clearer.

File: src/common.h (hunk 2 above)
Comment (SoniEx2) [reply src/common.h:143]: Shared with...?

File: src/wast-parser.cc [hunk 4]
```diff
@@ -2146,7 +2146,11 @@ Result WastParser::ParsePlainInstr(std::unique_ptr<Expr>* out_expr) {
           return Result::Error;
         }
 
+#if WABT_BIG_ENDIAN
```
Comment (SoniEx2) [inline src/wast-parser.cc:2149]: (side note: arguably this should just write the v128 directly, instead of doing this and then bitcasting. but it was done like this and there's probably a reason it was done like this but we personally don't know that reason and we don't think this makes sense. .-.)

File: src/common.h (hunk 2 above)
Comment (binji) [reply src/common.h:143]: It seems that this pattern is common in many of the changes below (using the negative offset for `WABT_BIG_ENDIAN` on memcpy). Can you make a function that does this, `MemcpyEndianAware` or some other better name?

Comment (SoniEx2) [reply src/common.h:143]: So like a MemcpyEndianAwareAt that takes "source object", "destination object", "source size", "destination size", "source offset", "destination offset", and "copy size"?

(this is a lot of arguments... should it be `inline`? where should we put it?)

Comment (binji) [reply src/common.h:143]: `common.h` seems like the best choice (maybe near `Bitcast`, etc). It is a lot of arguments, but I think that will be OK.

File: src/common.h [hunk 5]
```diff
@@ -101,6 +101,19 @@
 #define PRIaddress PRIu64
 #define PRIoffset PRIzx
 
+// Footgun: len may evaluate many times, and sizes may not be evaluated at all.
+#if WABT_BIG_ENDIAN
+#define WABT_MEMCPY_ENDIAN_AWARE(dst, src, dsize, ssize, doff, soff, len)               \
```
Comment (binji) [inline src/common.h:106]: Does this need to be a macro, or can it be an inline function instead?

File: src/common.h [hunk 6]
```diff
@@ -456,6 +461,15 @@ inline void ConvertBackslashToSlash(std::string* s) {
   ConvertBackslashToSlash(s->begin(), s->end());
 }
 
+inline void SwapBytesSized(void *addr, size_t size) {
+  auto bytes = static_cast<uint8_t*>(addr);
+  for (size_t i = 0; i < (size >> 1); i++) {
+    uint8_t tmp = bytes[i];
```
Comment (binji) [inline src/common.h:467]: use `std::swap()` instead

File: src/common.h [hunk 7]
```diff
@@ -456,6 +461,15 @@ inline void ConvertBackslashToSlash(std::string* s) {
   ConvertBackslashToSlash(s->begin(), s->end());
 }
 
+inline void SwapBytesSized(void *addr, size_t size) {
+  auto bytes = static_cast<uint8_t*>(addr);
+  for (size_t i = 0; i < (size >> 1); i++) {
```
Comment (binji) [inline src/common.h:466]: nit: use `size / 2` instead for clarity

File: src/common.h (hunk 5 above)
Comment (SoniEx2) [reply src/common.h:106]: it could be an inline function? but namespaces? this needs to be used in v128 but v128 is not in the namespace.

Comment (SoniEx2) [reply src/common.h:106]: (sorry, we don't really know C++. altho this does avoid evaluating `data_.size()` at all on little-endian? but compiler optimizations would probably optimize that out anyway...)

**Comment by SoniEx2** [comment]
Now this is pretty much just blocked on figuring out the C API stuff, either adjusting the C API tests to take memory order into account or making the C API copy stuff in the correct order into test-owned buffers and stuff. But at least the interpreter itself and wasm2c work.

File: wasm2c/wasm-rt-impl.c [hunk 8]
```diff
@@ -165,7 +165,14 @@ uint32_t wasm_rt_grow_memory(wasm_rt_memory_t* memory, uint32_t delta) {
   if (new_data == NULL) {
     return (uint32_t)-1;
   }
+#if WABT_BIG_ENDIAN
```
Comment (binji) [inline wasm2c/wasm-rt-impl.c:168]: `#if !WABT_BIG_ENDIAN` instead of using else block?

File: src/wast-parser.cc (hunk 4 above)
Comment (binji) [reply src/wast-parser.cc:2149]: If you switch this over to writing to the v128 directly, do any tests fail? I agree that this looks a little weird as it is.

**Review by binji — APPROVED** [review]
A few small comments, but otherwise lgtm. Sorry for the long review, and thanks for sticking with it!

**Comment by SoniEx2** [comment]
okay so we figured out what about the opcodecnt test is failing.

we're not fixing that. :smile: 

tl;dr: it has to do with this: https://github.com/WebAssembly/wabt/blob/master/src/binary-reader-opcnt.cc#L140-L181

```
>>> struct.pack("<f", 1.0)
b'\x00\x00\x80?'
>>> struct.pack("<f", 2.0)
b'\x00\x00\x00@'
>>> struct.pack(">f", 1.0)
b'?\x80\x00\x00'
>>> struct.pack(">f", 2.0)
b'@\x00\x00\x00'
```

**Comment by binji** [comment]
> okay so we figured out what about the opcodecnt test is failing.
> we're not fixing that. smile edit: we "fixed" it. :p

Oh, nasty :) Thanks for the fix.

File: src/prebuilt/wasm2c.include.c [hunk 9]
```diff
@@ -32,6 +32,37 @@ const char SECTION_NAME(declarations)[] =
...
+"    dest_chars[i] = dest_chars[n - i - 1];\n"
+"    dest_chars[n - i - 1] = cursor;\n"
+"  }\n"
+"}\n"
+"#define LOAD_DATA(m, o, i, s) load_data(&(m.data[m.size - o - s]), i, s)\n"
+"#define DEFINE_LOAD(name, t1, t2, t3)                                                 \\\n"
+"  static inline t3 name(wasm_rt_memory_t* mem, u64 addr) {                            \\\n"
+"    MEMCHECK(mem, addr, t1);                                                          \\\n"
+"    t1 result;                                                                        \\\n"
+"    __builtin_memcpy(&result, &mem->data[mem->size - addr - sizeof(t1)], sizeof(t1)); \\\n"
```
Comment (turbolent) [inline src/prebuilt/wasm2c.include.c:51]: @SoniEx2 @binji Maybe I'm missing something here, but is there any particular reason why the memory is addressed like this, instead of addressing it with `&mem->data[addr]` like in the little-endian?

Also, related, is there any way to make the `load_*` functions accessible?

Comment (SoniEx2) [reply src/prebuilt/wasm2c.include.c:51]: This avoids a byteswap. Whether it has any effects on performance on s390x specifically is questionable, but it can make a difference on other big-endian platforms.

And, the `load_*` functions?
